ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java

```diff
@@ -488,6 +489,230 @@ public void replicaThreadTestConverter() throws Exception {
+        addTtlUpdateMessagesToReplicasOfPartition(pid, ids.get(i), hostList);
+      }
+      // delete one of the keys that has put and ttl update on local host
+      addDeleteMessagesToReplicasOfPartition(pid, ids.get(1), remoteHostAndExpectedLocalHost);
+      // delete one of the keys that has put and ttl update on remote only
+      addDeleteMessagesToReplicasOfPartition(pid, ids.get(3), remoteHostOnly);
+
+      // add a TTL update and delete message without a put msg (compaction can create such a situation)
+      BlobId id = generateRandomBlobId(pid);
+      addTtlUpdateMessagesToReplicasOfPartition(pid, id, remoteHostOnly);
+      addDeleteMessagesToReplicasOfPartition(pid, id, remoteHostOnly);
+
+      // message transformation test cases
+      // a blob ID with PUT and TTL update in both remote and local
+      BlobId b0 = generateRandomBlobId(pid);
+      BlobId b0p = generateRandomBlobId(pid);
+      // a blob ID with a PUT in the local and PUT and TTL update in remote (with mapping)
+      BlobId b1 = generateRandomBlobId(pid);
+      BlobId b1p = generateRandomBlobId(pid);
+      // a blob ID with PUT and TTL update in remote only (with mapping)
+      BlobId b2 = generateRandomBlobId(pid);
+      BlobId b2p = generateRandomBlobId(pid);
+      // a blob ID with PUT and TTL update in remote (no mapping)
+      BlobId b3 = generateRandomBlobId(pid);
+      conversionMap.put(b0, b0p);
+      conversionMap.put(b1, b1p);
+      conversionMap.put(b2, b2p);
+      conversionMap.put(b3, null);
+
+      storeKeyConverter.convert(conversionMap.keySet());
+      // add as required on local, remote and expected local
+      // only PUT of b0p and b1p on local
+      addPutMessagesToReplicasOfPartition(Arrays.asList(b0p, b1p), localHostAndExpectedLocalHost);
+      // PUT of b0,b1,b2,b3 on remote
+      addPutMessagesToReplicasOfPartition(Arrays.asList(b0, b1, b2, b3), remoteHostOnly);
+      // PUT of b0, b1, b2 expected in local at the end
+      addPutMessagesToReplicasOfPartition(Collections.singletonList(b2), Collections.singletonList(transformer),
+          expectedLocalHostOnly);
+
+      // TTL update of b0 on all hosts
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b0p, localHostAndExpectedLocalHost);
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b0, remoteHostOnly);
+      // TTL update on b1, b2 and b3 on remote
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b1, remoteHostOnly);
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b1p, expectedLocalHostOnly);
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b2, remoteHostOnly);
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b2p, expectedLocalHostOnly);
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b3, remoteHostOnly);
+
+      numMessagesInEachPart = remoteHost.infosByPartition.get(pid).size();
+    }
+
+    Properties properties = new Properties();
```

Comment:
That's a good idea. I'll add that as a last commit before this code is merged (or we can add it after the merge). I don't want to make the change now because it will make the diff larger

Reply:
Addressed this